File: v87/work/geom/phase_bell.c
/*  phase_bell.c — v87 B1: in-kernel Bell construction from the fabric's phase
 *  structure. C reference implementation, and a race against phase_bell.py.
 *
 *  PROTOCOL (identical physics to phase_bell.py):
 *    object 1 phase  lambda + omega t          -> wing A, detector offset a
 *    object 2 phase  lambda + pi + omega t     -> wing B, detector offset b
 *    detectors carry the SAME omega, so omega*t cancels in both relative phases
 *    readout (A6 fuse/repel dichotomy):  A = sgn cos(lambda - a)
 *                                        B = sgn cos(lambda + pi - b)
 *    arm 1: lambda ~ Uniform                       -- (MI) respected
 *    arm 2: lambda ~ |cos(lambda - b)|             -- (MI) violated, p = 1 tilt
 *
 *  THREE ALGORITHMS, so the race separates language from algorithm:
 *    NAIVE  O(grid^3 * N)   — what the Python does: re-sum all N samples for
 *                             every one of grid^3 setting triples.
 *    TABLE  O(grid*N + grid^3) — note E(a,b) depends only on (a-b): substituting
 *                             u = lambda - b makes the weight and B functions of
 *                             u alone and A a function of u + b - a. So one
 *                             correlation table E(theta) serves every triple.
 *    HIST   O(N + nb^2 + grid^3) — histogram lambda once, then the whole E(theta)
 *                             table is a circular correlation of two square waves
 *                             against that histogram. The N-dependence collapses
 *                             to a single pass.
 *
 *  Build: gcc -O3 -march=native -fopenmp -o phase_bell phase_bell.c -lm
 *  Usage: phase_bell [N] [grid] [--race]
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <stdint.h>
#include <omp.h>

#define TAU 6.283185307179586476925287

/* xoshiro256++ — fast, good, and reproducible across runs */
static inline uint64_t rotl(const uint64_t x, int k) { return (x << k) | (x >> (64 - k)); }
typedef struct { uint64_t s[4]; } rng_t;
static inline uint64_t rnext(rng_t *r) {
    const uint64_t res = rotl(r->s[0] + r->s[3], 23) + r->s[0];
    const uint64_t t = r->s[1] << 17;
    r->s[2] ^= r->s[0]; r->s[3] ^= r->s[1]; r->s[1] ^= r->s[2]; r->s[0] ^= r->s[3];
    r->s[2] ^= t; r->s[3] = rotl(r->s[3], 45);
    return res;
}
static void rseed(rng_t *r, uint64_t seed) {
    for (int i = 0; i < 4; i++) {
        seed += 0x9E3779B97F4A7C15ULL;
        uint64_t z = seed;
        z = (z ^ (z >> 30)) * 0xBF58476D1CE4E5B9ULL;
        z = (z ^ (z >> 27)) * 0x94D049BB133111EBULL;
        r->s[i] = z ^ (z >> 31);
    }
}
static inline double runif(rng_t *r) { return (rnext(r) >> 11) * 0x1.0p-53; }

static double now(void) { return omp_get_wtime(); }

/* ---------------------------------------------------------------- NAIVE */
/* exactly the Python algorithm: for every (a',b,b') resum all N samples */
static double chsh_naive(const double *lam, long N, int grid, int arm) {
    double best = 0.0;
    #pragma omp parallel for schedule(dynamic) reduction(max:best)
    for (int ia = 0; ia < grid; ia++) {
        double ap = TAU * ia / grid;
        for (int ib = 0; ib < grid; ib++) {
            double b = TAU * ib / grid;
            for (int ic = 0; ic < grid; ic++) {
                double bp = TAU * ic / grid;
                double e[4] = {0, 0, 0, 0};
                double wsum[2] = {0, 0};
                for (long i = 0; i < N; i++) {
                    double L = lam[i];
                    double sb = (cos(L - b) >= 0) ? -1.0 : 1.0;    /* B at b (with the pi) */
                    double sbp = (cos(L - bp) >= 0) ? -1.0 : 1.0;
                    double s0 = (cos(L) >= 0) ? 1.0 : -1.0;        /* A at a = 0 */
                    double sa = (cos(L - ap) >= 0) ? 1.0 : -1.0;   /* A at a' */
                    double wb = (arm == 2) ? fabs(cos(L - b)) : 1.0;
                    double wbp = (arm == 2) ? fabs(cos(L - bp)) : 1.0;
                    e[0] += wb * s0 * sb;  e[1] += wbp * s0 * sbp;
                    e[2] += wb * sa * sb;  e[3] += wbp * sa * sbp;
                    wsum[0] += wb; wsum[1] += wbp;
                }
                double S = e[0] / wsum[0] + e[1] / wsum[1]
                         + e[2] / wsum[0] - e[3] / wsum[1];
                if (fabs(S) > best) best = fabs(S);
            }
        }
    }
    return best;
}

/* ---------------------------------------------------------------- TABLE */
/* E depends only on theta = a - b: build one table, then search by lookup */
static void build_table_direct(const double *lam, long N, int nth, int arm, double *E) {
    #pragma omp parallel for schedule(static)
    for (int k = 0; k < nth; k++) {
        double th = TAU * k / nth;
        double num = 0.0, den = 0.0;
        for (long i = 0; i < N; i++) {
            double u = lam[i];                       /* u = lambda - b */
            double w = (arm == 2) ? fabs(cos(u)) : 1.0;
            double B = (cos(u) >= 0) ? -1.0 : 1.0;   /* includes the pi offset */
            double A = (cos(u + th) >= 0) ? 1.0 : -1.0;
            num += w * A * B; den += w;
        }
        E[k] = num / den;
    }
}

/* ---------------------------------------------------------------- HIST */
/* one O(N) pass to histogram lambda; then the table is a circular correlation */
static void build_table_hist(const double *lam, long N, int nb, int nth, int arm,
                             double *E) {
    double *h = calloc(nb, sizeof(double));
    #pragma omp parallel
    {
        double *hl = calloc(nb, sizeof(double));
        #pragma omp for schedule(static)
        for (long i = 0; i < N; i++) {
            double u = fmod(lam[i], TAU); if (u < 0) u += TAU;
            int j = (int)(u * nb / TAU); if (j >= nb) j = nb - 1;
            hl[j] += (arm == 2) ? fabs(cos(u)) : 1.0;
        }
        #pragma omp critical
        for (int j = 0; j < nb; j++) h[j] += hl[j];
        free(hl);
    }
    double den = 0.0;
    for (int j = 0; j < nb; j++) den += h[j];
    /* B(u) and A(u+theta) are square waves; correlate against h */
    signed char *Bs = malloc(nb), *As = malloc(nb);
    for (int j = 0; j < nb; j++) {
        double u = TAU * (j + 0.5) / nb;
        Bs[j] = (cos(u) >= 0) ? -1 : 1;
        As[j] = (cos(u) >= 0) ? 1 : -1;
    }
    #pragma omp parallel for schedule(static)
    for (int k = 0; k < nth; k++) {
        int shift = (int)llround((double)k * nb / nth);
        double num = 0.0;
        for (int j = 0; j < nb; j++) {
            int jj = j + shift; if (jj >= nb) jj -= nb;
            num += h[j] * (double)As[jj] * (double)Bs[j];
        }
        E[k] = num / den;
    }
    free(h); free(Bs); free(As);
}

/* ---------------------------------------------------------------- EXACT INT
 * Angles as uint32 "turns": 2^32 == one full revolution. Modular angle
 * arithmetic is then EXACT -- no fmod, no representation of pi, and the
 * response square waves need no cos() call at all:
 *     sgn cos(u) = +1  iff  turn in [0, 1/4) U [3/4, 1)
 * For arm 1 the histogram bins are integer COUNTS and the correlation is an
 * integer sum, so E = (integer)/N is computed with ZERO rounding.
 * For arm 2 the weights are irrational; they are taken to Q30 fixed point and
 * accumulated in __int128, which removes overflow as a concern entirely.   */
static void build_table_int(const double *lam, long N, int nb, int nth, int arm,
                            double *E, long *exact_num, long *exact_den) {
    __int128 *h = calloc(nb, sizeof(__int128));
    const int SH = 30;                       /* Q30 fixed point for arm 2 */
    long long *wq = malloc(sizeof(long long) * nb);
    for (int j = 0; j < nb; j++) {
        double u = TAU * (j + 0.5) / nb;
        wq[j] = (arm == 2) ? (long long)llround(fabs(cos(u)) * (1LL << SH))
                           : 1LL;
    }
    #pragma omp parallel
    {
        __int128 *hl = calloc(nb, sizeof(__int128));
        #pragma omp for schedule(static)
        for (long i = 0; i < N; i++) {
            /* exact modular reduction into a uint32 turn */
            double frac = lam[i] / TAU;
            frac -= floor(frac);
            uint32_t turn = (uint32_t)(frac * 4294967296.0);
            int j = (int)(turn >> (32 - (int)llround(log2((double)nb))));
            hl[j] += wq[j];
        }
        #pragma omp critical
        for (int j = 0; j < nb; j++) h[j] += hl[j];
        free(hl);
    }
    __int128 den = 0;
    for (int j = 0; j < nb; j++) den += h[j];
    /* square waves without cos(): +1 on the first and last quarter */
    signed char *As = malloc(nb);
    for (int j = 0; j < nb; j++)
        As[j] = (j < nb / 4 || j >= 3 * nb / 4) ? 1 : -1;
    #pragma omp parallel for schedule(static)
    for (int k = 0; k < nth; k++) {
        int shift = (int)llround((double)k * nb / nth);
        __int128 num = 0;
        for (int j = 0; j < nb; j++) {
            int jj = j + shift; if (jj >= nb) jj -= nb;
            /* B = -As[j] (the pi offset), A = As[jj] */
            num += (__int128)h[j] * (__int128)(-(int)As[jj] * (int)As[j]);
        }
        E[k] = (double)num / (double)den;
        if (k == 0 && exact_num) { *exact_num = (long)num; *exact_den = (long)den; }
    }
    free(h); free(As); free(wq);
}

static double chsh_from_table(const double *E, int nth, int grid) {
    double best = 0.0;
    #pragma omp parallel for schedule(static) reduction(max:best)
    for (int ia = 0; ia < grid; ia++) {
        double ap = TAU * ia / grid;
        for (int ib = 0; ib < grid; ib++) {
            double b = TAU * ib / grid;
            for (int ic = 0; ic < grid; ic++) {
                double bp = TAU * ic / grid;
                #define ELOOK(th) E[((int)llround(fmod((th) + TAU * 4, TAU) * nth / TAU)) % nth]
                double S = ELOOK(0.0 - b) + ELOOK(0.0 - bp)
                         + ELOOK(ap - b) - ELOOK(ap - bp);
                if (fabs(S) > best) best = fabs(S);
            }
        }
    }
    return best;
}

int main(int argc, char **argv) {
    long N = (argc > 1) ? atol(argv[1]) : 4000000L;
    int grid = (argc > 2) ? atoi(argv[2]) : 49;
    int race = 0;
    for (int i = 1; i < argc; i++) if (!strcmp(argv[i], "--race")) race = 1;

    printf("phase_bell (C)  N=%ld  grid=%d  threads=%d\n", N, grid, omp_get_max_threads());

    double *lam = malloc(sizeof(double) * N);
    rng_t r; rseed(&r, 20260726ULL);
    for (long i = 0; i < N; i++) lam[i] = runif(&r) * TAU;

    int nth = 2880, nb = 4096;
    for (int i = 1; i < argc; i++) {
        if (!strcmp(argv[i], "--nb") && i + 1 < argc) nb = atoi(argv[i + 1]);
        if (!strcmp(argv[i], "--nth") && i + 1 < argc) nth = atoi(argv[i + 1]);
    }
    double *E = malloc(sizeof(double) * nth);

    for (int arm = 1; arm <= 2; arm++) {
        printf("\n--- ARM %d (%s) ---\n", arm,
               arm == 1 ? "MI respected, uniform creation phase"
                        : "MI violated by design, p=1 tilt");
        double t0, S;

        long en = 0, ed = 0;
        t0 = now(); build_table_int(lam, N, nb, nth, arm, E, &en, &ed);
        double Si = chsh_from_table(E, nth, grid); double th_i = now() - t0;
        printf("  INT    max|S| = %.9f  %8.3f s   [exact E(0) = %ld/%ld = %.15f]\n",
               Si, th_i, en, ed, (double)en / (double)ed);

        t0 = now(); build_table_hist(lam, N, nb, nth, arm, E);
        double Sh = chsh_from_table(E, nth, grid); double th_h = now() - t0;
        printf("  HIST   max|S| = %.9f  %8.3f s   [nb=%d nth=%d]   f64-vs-int diff %.2e\n",
               Sh, th_h, nb, nth, fabs(Sh - Si));

        if (race) {
            t0 = now(); build_table_direct(lam, N, nth, arm, E);
            double Sd = chsh_from_table(E, nth, grid); double th_d = now() - t0;
            printf("  TABLE  max|S| = %.5f      %8.3f s\n", Sd, th_d);
        }

        if (race) {
            t0 = now(); S = chsh_naive(lam, N, grid, arm); double th_n = now() - t0;
            printf("  NAIVE  max|S| = %.5f      %8.3f s   (%.0fx slower than HIST)\n",
                   S, th_n, th_n / th_h);
        }
        printf("  E(theta) samples: ");
        for (int d = 0; d <= 180; d += 30)
            printf("%+.4f ", E[(int)llround((double)d / 360.0 * nth)]);
        printf("\n");
    }
    printf("\n  targets: arm1 triangle -> 2.00000 ; arm2 cosine -> 2sqrt2 = %.5f\n",
           2.0 * sqrt(2.0));
    free(lam); free(E);
    return 0;
}

File: v87/work/geom/phase_bell.py
#!/usr/bin/env python3
"""v87 B1 — formal construction of an in-kernel Bell test from the fabric's own
phase structure, and its numerical demonstration.

THE IDEA (crank 4). For a stationary fabric object Phi_a = f(r) e^{i omega t},
the profile f(r) is static: the ONLY thing that evolves is the phase. Time IS
phase advance. Two consequences chain:

  (i)  For two IDENTICAL objects (same omega, hence same Q by the branch
       relation Q = f(omega)), the relative phase is FROZEN: both advance at
       omega, so Delta phi(t) = Delta phi(0) for all t. A conserved, transport-
       surviving shared hidden variable.

  (ii) A detector is itself a fabric object (O2). If it carries the SAME omega,
       then in the relative phase the omega*t CANCELS:
              theta_A = (lambda + omega t) - (a + omega t) = lambda - a
       so the measured angle is time-independent. The setting is the detector's
       phase OFFSET a, which is freely dialable at fixed Q. (A setting cannot be
       a phase RATE: changing omega changes Q via the branch, i.e. changes the
       detector into a different object.)

Together these supply all four ingredients a Bell test needs, which the crank-2
review listed as gap #1 "what is a setting in the kernel? undefined".

This file constructs the protocol explicitly and measures what it gives.
"""
import numpy as np

TAU = 2 * np.pi
rng = np.random.default_rng(20260726)
N = 4_000_000


# ---------------------------------------------------------------- the protocol
def outcomes(lam, a, b, Delta0=np.pi):
    """Local, deterministic, dichotomic readouts.

    Object 1 phase: lam + omega t          -> wing A, detector offset a
    Object 2 phase: lam + Delta0 + omega t -> wing B, detector offset b
    Detectors carry the same omega, so omega*t cancels in both relative phases.

    Readout is the A6 fuse/repel dichotomy: co-phase (|rel| < pi/2) -> +1,
    anti-phase -> -1, i.e. sgn(cos(relative phase)).
    """
    A = np.sign(np.cos(lam - a))
    B = np.sign(np.cos(lam + Delta0 - b))
    return A, B


def E_corr(lam, a, b, w=None):
    A, B = outcomes(lam, a, b)
    if w is None:
        return float(np.mean(A * B))
    return float(np.average(A * B, weights=w))


def chsh(lam_fn, coarse=97):
    """max |S| over settings, where lam_fn(b) returns the lambda sample that
    is realised when Bob's setting is b (this is where (MI) lives)."""
    best, arg = 0.0, None
    grid = np.linspace(0, TAU, coarse, endpoint=False)
    for ap in grid:
        for b in grid:
            for bp in grid:
                lb, wb = lam_fn(b)
                lbp, wbp = lam_fn(bp)
                S = (E_corr(lb, 0.0, b, wb) + E_corr(lbp, 0.0, bp, wbp)
                     + E_corr(lb, ap, b, wb) - E_corr(lbp, ap, bp, wbp))
                if abs(S) > best:
                    best, arg = abs(S), (0.0, ap, b, bp)
    return best, arg


# ---------------------------------------------------------------- checks
def check_time_cancellation():
    print("=" * 74)
    print("1.  The omega*t cancellation -- the measured angle is time-independent")
    print("=" * 74)
    omega = 1.4300                      # v86 branch clock
    lam0 = 0.7
    a, b = 0.3, 1.1
    print("   omega = %.4f, lambda(0) = %.3f, a = %.3f, b = %.3f" % (omega, lam0, a, b))
    print("   %10s %14s %14s %8s %8s" % ("t", "obj1 phase", "detA phase", "rel", "A"))
    for t in (0.0, 1.0, 7.3, 55.0, 1000.0):
        ph_obj = (lam0 + omega * t) % TAU
        ph_det = (a + omega * t) % TAU
        rel = (ph_obj - ph_det) % TAU
        A = np.sign(np.cos(rel))
        print("   %10.1f %14.5f %14.5f %8.4f %8.0f" % (t, ph_obj, ph_det, rel, A))
    print("   -> relative phase constant at lambda - a = %.4f for all t.\n"
          % ((lam0 - a) % TAU))
    print("   Same-omega detectors are REQUIRED. With a detuned detector the")
    print("   relative phase drifts at (omega_obj - omega_det) t, and the outcome")
    print("   depends on WHEN the interaction happens -- reintroducing timing as")
    print("   an uncontrolled variable. Q = f(omega) then forces the detector to")
    print("   carry the same CHARGE as the object: the analyser is the same")
    print("   species as the thing analysed, which is O2-consistent.\n")


def check_frozen_relative_phase():
    print("=" * 74)
    print("2.  Frozen relative phase for identical objects")
    print("=" * 74)
    omega = 1.4300
    d0 = 1.234
    ts = np.array([0.0, 10.0, 100.0, 2000.0])
    drift = ((d0 + omega * ts) - (omega * ts)) % TAU - d0
    print("   identical pair (same omega): max |Delta phi(t) - Delta phi(0)| = %.2e"
          % np.max(np.abs(drift)))
    for dw in (1e-4, 1e-3, 1e-2):
        print("   detuned by Delta omega = %.0e : Delta phi drifts %.3f rad over "
              "T=2000  -> %s" % (dw, dw * 2000,
                                 "FROZEN" if dw * 2000 < 0.1 else "SCRAMBLED"))
    print("""
   This is the part that answers G14 in part. GROK's obstruction was that the
   kernel's mixing destroys any correlation prepared in the common past. For
   IDENTICAL monochromatic objects there is nothing to mix: both clocks tick at
   the same rate and the relative phase is a constant of the pair. The fabric
   already relies on this -- A6's co-phase-fuses / anti-phase-repels distinction
   requires a persistent Delta phi, and the v85 pair runs are built on it.
   Tolerance: the pair must be matched to |Delta omega| << 1/T.
""")


def arm1():
    print("=" * 74)
    print("3.  ARM 1 -- (MI) respected. Independent streams for lambda and settings.")
    print("=" * 74)
    def lam_fn(b):
        return rng.uniform(0, TAU, N), None     # lambda independent of b
    S, arg = chsh(lam_fn, coarse=49)
    print("   creation phase uncontrolled, settings from a separate stream")
    print("   max|S| = %.5f      (classical bound 2, triangle rung)" % S)
    ths = np.linspace(0, np.pi, 7)
    lam = rng.uniform(0, TAU, N)
    meas = [E_corr(lam, 0.0, np.pi - t) for t in ths]     # Delta0=pi folds in
    tri = [-1 + 2 * t / np.pi for t in ths]
    print("   E(theta):  " + " ".join("%+.4f" % v for v in meas))
    print("   triangle:  " + " ".join("%+.4f" % v for v in tri))
    print("   max dev from the triangle wave = %.4f"
          % max(abs(m - t) for m, t in zip(meas, tri)))
    print("\n   PREDICTION for the kernel: S = 2.00, not merely S <= 2.\n")
    return S


def arm2():
    print("=" * 74)
    print("4.  ARM 2 -- (MI) violated BY DESIGN. Can the fabric HOST Door A?")
    print("=" * 74)
    def lam_fn(b):
        lam = rng.uniform(0, TAU, N)
        w = np.abs(np.cos(lam + np.pi - b))     # p = 1 tilt on the B-side angle
        return lam, w
    S, arg = chsh(lam_fn, coarse=49)
    print("   creation phase drawn with the p=1 tilt rho(lambda|b) ~ |cos(lambda-b)|")
    print("   max|S| = %.5f      (Tsirelson 2sqrt2 = %.5f)" % (S, 2 * np.sqrt(2)))

    # cost of the tilt, measured
    nb, nl = 2, 64
    bs = [0.0, np.pi / 2]
    joint = np.zeros((nb, nl))
    for i, b in enumerate(bs):
        lam = rng.uniform(0, TAU, N // 4)
        w = np.abs(np.cos(lam + np.pi - b))
        h, _ = np.histogram(lam, bins=nl, range=(0, TAU), weights=w)
        joint[i] = h / h.sum() * 0.5
    pl = joint.sum(axis=0)
    pb = joint.sum(axis=1)
    I = 0.0
    for i in range(nb):
        for j in range(nl):
            if joint[i, j] > 0:
                I += joint[i, j] * np.log2(joint[i, j] / (pb[i] * pl[j]))
    print("   measured cost I(lambda ; b) = %.4f bits" % I)
    print("""
   NOTE the honest comparison: GROK's information-OPTIMAL cost for CHSH at
   Tsirelson is 0.046274 bits. The |cos| tilt used here is the EXACT-singlet
   construction, not the information-optimal one, so it costs more. Both reach
   2sqrt2; they differ in efficiency, not in reachability.

   What arm 2 establishes: the fabric's phase structure can CARRY the Door A
   mechanism. What it does NOT establish: that fabric dynamics GENERATE the
   tilt. The tilt is imposed here by drawing the creation phase from a
   setting-dependent distribution -- exactly the thing that is postulated, not
   derived.
""")
    return S, I


def arm3_readout_fuzziness():
    print("=" * 74)
    print("5.  ARM 3 -- readout fuzziness is a DETECTION LOOPHOLE, not noise")
    print("=" * 74)
    print("""
The A6 fuse/repel dichotomy is sharp only away from |relative phase| = pi/2.
Near the boundary the dynamical outcome is marginal. If marginal events are
DISCARDED rather than assigned, the surviving sub-ensemble is biased toward
well-aligned configurations -- which is precisely crank-2 Reading 2, already
refuted on efficiency grounds. Quantifying the damage:
""")
    lam = rng.uniform(0, TAU, N)
    print("   %10s %10s %12s" % ("dead zone", "efficiency", "max|S| (MI respected)"))
    for dz in (0.0, 0.10, 0.25, 0.50):
        def lam_fn(b, dz=dz):
            keep = np.abs(np.cos(lam - b)) > np.sin(dz)
            return lam[keep], None
        # efficiency
        eff = float(np.mean(np.abs(np.cos(lam)) > np.sin(dz)))
        S, _ = chsh(lam_fn, coarse=25)
        print("   %10.2f %10.4f %12.5f" % (dz, eff, S))
    print("""
   Discarding marginal events inflates S above 2 with NO measurement dependence
   at all -- a pure post-selection artifact. Any in-kernel run MUST assign every
   event (e.g. by the sign of the residual interaction, or a deterministic
   tie-break), and MUST report the fraction of marginal events. An S > 2 obtained
   with discards is the closed detection loophole and must be rejected.
""")


if __name__ == "__main__":
    check_time_cancellation()
    check_frozen_relative_phase()
    s1 = arm1()
    s2, I = arm2()
    arm3_readout_fuzziness()
    print("=" * 74)
    print("SUMMARY")
    print("=" * 74)
    print("   arm 1 (MI respected, fabric's own uniform creation phase) : S = %.4f" % s1)
    print("   arm 2 (MI violated by design, p=1 tilt)                   : S = %.4f "
          "at %.3f bits" % (s2, I))
    print("   arm 3 (marginal events discarded)                         : see above")
